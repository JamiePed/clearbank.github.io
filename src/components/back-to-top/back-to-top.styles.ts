import styled, { css } from 'styled-components'

import { colors } from 'src/components/theme'
import {
  animationButtonIn,
  animationButtonOut
} from 'src/components/clearbank-ui/styles'

const transitionDuration = '0.5s'

export const Container = styled.div`
  position: fixed;
  bottom: 40px;
  right: 10px;
  will-change: visibility, opacity, transition;
  cursor: default;
  visibility: hidden;
  opacity: 0;

  // wait transitionDuration before changing visibility since visibility is not animatable
  transition: visibility 0s linear ${transitionDuration},
    opacity ${transitionDuration};

  ${({ isVisible }) =>
    isVisible &&
    css`
      cursor: pointer;
      visibility: visible;
      opacity: 1;
      // change visibility right away then transition opacity
      transition: visibility 0s linear 0s, opacity ${transitionDuration};
    `}
`

export const Button = styled.button`
  width: 60px;
  height: 60px;
  border: none;
  border-radius: 50%;
  text-align: center;
  box-shadow: 0 5px 10px 0 rgba(73, 151, 222, 0.2);
  background: ${colors.brandLight};
  color: ${colors.ceruleanBlue};
  overflow: hidden;
  outline: none;
  cursor: inherit;
  transition: all ${transitionDuration};

  &:hover {
    background: ${colors.brandSecondaryDarkest};
    color: ${colors.brandLight};
  }

  &:focus,
  &:active {
    background: ${colors.brandInfoDark};
    color: ${colors.brandLight};
  }
`

export const Icon = styled.i`
  display: block;
  font-size: 24px;

  // hover on parent
  ${Button}:hover & {
    animation: ${animationButtonIn} 0.15s ease-in,
      ${animationButtonOut} 0.15s 0.15s ease-out;
  }
`
