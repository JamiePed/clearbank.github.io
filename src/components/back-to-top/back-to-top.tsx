import React, { useState, useEffect } from 'react'
import { hasWindow } from 'src/utils/browser.services'
import throttle from 'lodash.throttle'

import * as Types from './back-to-top.types'
import * as Styles from './back-to-top.styles'

const BackToTop: React.FunctionComponent<Types.BackToTopProps> = (): JSX.Element => {
  const [isVisible, setIsVisible] = useState(false)

  function handleClick () {
    if (!hasWindow) {
      return
    }

    window.scroll({
      top: 0,
      left: 0,
      behavior: 'smooth'
    })
  }

  useEffect(() => {
    const handleScroll = () => {
      const currentScrollPosition = window.scrollY
      const visibilityThreshold = 500

      setIsVisible(currentScrollPosition >= visibilityThreshold)
    }

    const throttledScroll = throttle(handleScroll, 300)

    window.addEventListener('scroll', throttledScroll)

    return () => {
      window.removeEventListener('scroll', throttledScroll)
    }
  }, [])

  return (
    <Styles.Container isVisible={isVisible}>
      <Styles.Button onClick={handleClick}>
        <Styles.Icon className='cb-icon-arrow-up' />
      </Styles.Button>
    </Styles.Container>
  )
}

export default BackToTop
