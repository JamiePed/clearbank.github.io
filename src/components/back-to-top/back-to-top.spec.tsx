import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'

import Component from './back-to-top'

import * as Types from './back-to-top.types'

let component: ReturnType<typeof render> = null
let fragment: DocumentFragment = null

let root: HTMLElement = null

jest.mock('src/utils/browser.services', () => ({
  hasWindow: () => {
    return true
  }
}))

const defaultProps: Types.BackToTopProps = {}

beforeEach(() => {
  component = render(<Component {...defaultProps} children='children' />)

  fragment = component.asFragment()
  root = component.container.firstChild
})

afterEach(() => {
  fragment = null
  root = null
})

describe('Component', () => {
  test('matches snapshot', () => {
    expect(root).toMatchSnapshot()
  })

  test('has child elements', () => {
    expect(root.hasChildNodes()).toBeTruthy()
  })

  test('has correct icon', () => {
    expect(root.querySelectorAll('.cb-icon-arrow-up').length).toBeTruthy
    expect(root.querySelector('.cb-icon-arrow-up').tagName.toLowerCase()).toBe(
      'i'
    )
  })
})
