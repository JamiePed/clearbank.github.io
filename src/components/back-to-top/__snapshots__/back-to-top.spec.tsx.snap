// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component matches snapshot 1`] = `
.c0 {
  position: fixed;
  bottom: 40px;
  right: 10px;
  will-change: visibility,opacity,transition;
  cursor: default;
  visibility: hidden;
  opacity: 0;
  -webkit-transition: visibility 0s linear 0.5s, opacity 0.5s;
  transition: visibility 0s linear 0.5s, opacity 0.5s;
}

.c2 {
  width: 60px;
  height: 60px;
  border: none;
  border-radius: 50%;
  text-align: center;
  box-shadow: 0 5px 10px 0 rgba(73,151,222,0.2);
  background: #fff;
  color: #3048D1;
  overflow: hidden;
  outline: none;
  cursor: inherit;
  -webkit-transition: all 0.5s;
  transition: all 0.5s;
}

.c2:hover {
  background: #0D326E;
  color: #fff;
}

.c2:focus,
.c2:active {
  background: #0B2145;
  color: #fff;
}

.c3 {
  display: block;
  font-size: 24px;
}

.c1:hover .c3 {
  -webkit-animation: frYDOQ 0.15s ease-in, kjJcRz 0.15s 0.15s ease-out;
  animation: frYDOQ 0.15s ease-in, kjJcRz 0.15s 0.15s ease-out;
}

<div
  class="c0"
>
  <button
    class="c1 c2"
  >
    <i
      class="c3 cb-icon-arrow-up"
    />
  </button>
</div>
`;
